src/link.js

```diff
@@ -113,11 +95,29 @@ module.exports = function link(config, args, callback) {
   };
 
   const tasks = dependencies.map((dependency) => (next) => {
-    const prelink = makeHook(dependency, 'prelink');
-    const postlink = makeHook(dependency, 'postlink');
-    const link = makeLink(dependency);
+    const queue = [];
+
+    var prelink;
+    var postlink;
+
+    if (dependency.config.hooks) {
+      prelink = dependency.config.hooks.prelink;
+      postlink = dependency.config.hooks.postlink;
+    }
+
+    if (prelink) {
+      queue.push(makeHook(prelink));
+    }
+
+    queue.push(makeLink(dependency));
+
+    if (postlink) {
+      queue.push(makeHook(postlink));
+    }
+
+    queue.push(next);
 
-    async.waterfall([prelink, link, postlink, next]);
+    async.waterfall(queue);
```

Comment:
There's no `callback` to waterfall. I think it should be `async.waterfall([...], next)` instead.